### PR TITLE
Docker

### LINKED ISSUE #14: artwork service - add docker
containerize the service with docker

### PR DESCRIPTION
**Pull request by JordanCaseCode**
Adding docker compose to stand up service and mongo separately 

Fixes #14

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+# create an argument in the docker image for the jar file
+ARG JAR_FILE=target/artwork-0.0.1-SNAPSHOT.jar
+
+# cd /opt/app
+WORKDIR /opt/app
+
+# cp target/artwork-0.0.1-SNAPSHOT.jar /opt/app/app.jar
+COPY ${JAR_FILE} app.jar
+
```
Comment (cgrountree) [inline Dockerfile:12]: can you remove the extra line here 😂

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+# expose the port 8080 for the service
+EXPOSE 8080
+
```
Comment (cgrountree) [inline Dockerfile]: same as above

**Review by cgrountree — CHANGES_REQUESTED** [review]
when running `mvn clean isntall` im seeing the following warning:

```
[WARNING]
[WARNING] Some problems were encountered while building the effective model for com.dorksquad:artwork:jar:0.0.1-SNAPSHOT
[WARNING] 'dependencies.dependency.version' for org.jetbrains:annotations:jar is either LATEST or RELEASE (both of them are being deprecated) @ line 67, column 22
[WARNING]
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING]
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING]
```

not sure how soon this will be deprecated :thinking: 


i'm also seeing the following log right before the tests finishing running:

```
2021-07-17 18:01:01.390  INFO 8511 --- [localhost:49157] org.mongodb.driver.cluster               : Exception in monitor thread while connecting to server localhost:49157

com.mongodb.MongoSocketReadException: Prematurely reached end of stream
	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:112) ~[mongodb-driver-core-4.1.2.jar:na]
	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:131) ~[mongodb-driver-core-4.1.2.jar:na]
	at com.mongodb.internal.connection.InternalStreamConnection.receiveResponseBuffers(InternalStreamConnection.java:648) ~[mongodb-driver-core-4.1.2.jar:na]
	at com.mongodb.internal.connection.InternalStreamConnection.receiveMessageWithAdditionalTimeout(InternalStreamConnection.java:513) ~[mongodb-driver-core-4.1.2.jar:na]
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:356) ~[mongodb-driver-core-4.1.2.jar:na]
	at com.mongodb.internal.connection.InternalStreamConnection.receive(InternalStreamConnection.java:316) ~[mongodb-driver-core-4.1.2.jar:na]
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.lookupServerDescription(DefaultServerMonitor.java:215) ~[mongodb-driver-core-4.1.2.jar:na]
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.run(DefaultServerMonitor.java:144) ~[mongodb-driver-core-4.1.2.jar:na]
	at java.base/java.lang.Thread.run(Thread.java:831) ~[na:na]
```
the tests pass, but im not sure if this is expected or not

File: Dockerfile (hunk 1 above)
Comment (JordanCaseCode) [reply Dockerfile:12]: no

File: Dockerfile (hunk 2 above)
Comment (JordanCaseCode) [reply Dockerfile]: never

**Comment by JordanCaseCode** [comment]
> when running `mvn clean isntall` im seeing the following warning:
> 
> ```
> [WARNING]
> [WARNING] Some problems were encountered while building the effective model for com.dorksquad:artwork:jar:0.0.1-SNAPSHOT
> [WARNING] 'dependencies.dependency.version' for org.jetbrains:annotations:jar is either LATEST or RELEASE (both of them are being deprecated) @ line 67, column 22
> [WARNING]
> [WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
> [WARNING]
> [WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
> [WARNING]
> ```
> 
> not sure how soon this will be deprecated 
> 
> i'm also seeing the following log right before the tests finishing running:
> 
> ```
> 2021-07-17 18:01:01.390  INFO 8511 --- [localhost:49157] org.mongodb.driver.cluster               : Exception in monitor thread while connecting to server localhost:49157
> 
> com.mongodb.MongoSocketReadException: Prematurely reached end of stream
> 	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:112) ~[mongodb-driver-core-4.1.2.jar:na]
> 	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:131) ~[mongodb-driver-core-4.1.2.jar:na]
> 	at com.mongodb.internal.connection.InternalStreamConnection.receiveResponseBuffers(InternalStreamConnection.java:648) ~[mongodb-driver-core-4.1.2.jar:na]
> 	at com.mongodb.internal.connection.InternalStreamConnection.receiveMessageWithAdditionalTimeout(InternalStreamConnection.java:513) ~[mongodb-driver-core-4.1.2.jar:na]
> 	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:356) ~[mongodb-driver-core-4.1.2.jar:na]
> 	at com.mongodb.internal.connection.InternalStreamConnection.receive(InternalStreamConnection.java:316) ~[mongodb-driver-core-4.1.2.jar:na]
> 	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.lookupServerDescription(DefaultServerMonitor.java:215) ~[mongodb-driver-core-4.1.2.jar:na]
> 	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.run(DefaultServerMonitor.java:144) ~[mongodb-driver-core-4.1.2.jar:na]
> 	at java.base/java.lang.Thread.run(Thread.java:831) ~[na:na]
> ```
> 
> the tests pass, but im not sure if this is expected or not

Both are resolved.
1 - jetbrains version just needed to be set to a non-dynamic version (16.0.1 instead of RELEASE)
2 - I removed the closing of the mongodb docker container at the end of the test class. This is handled with autoCloseable. So it was basically just a harsh closing of the container while the testcontainer was still trying to handle it.

**Comment by cgrountree** [comment]
was able to confirm with this branch https://github.com/dorksquad/artwork/pull/19 that insertion into mongo via the addArtworks endpoint is successful before the addition of docker compose.

so we probably just need to continue trying to find how to get logging for the containers to track down exactly why the insertion doesn't happen here.

**Review by cgrountree — APPROVED** [review]
:boom:

**Pull request merged**